File: src/sdk/go/server_http.go
package appmesh

import (
	"errors"
	"log"
	"net/http"
	"net/url"
	"os"
	"time"
)

// AppMeshServerHttpContext interacts with the App Mesh REST service over HTTPS for server-side applications.
// Requires environment variables: APP_MESH_PROCESS_KEY and APP_MESH_APPLICATION_NAME.
type AppMeshServerHttpContext struct {
	client *AppMeshClient
}

// NewHTTPContext creates a server-side task context backed by the HTTP client.
// Uses default HTTPS endpoint and TLS settings unless overridden via Option.
func NewHTTPContext(options Option) (*AppMeshServerHttpContext, error) {
	httpClient, err := NewHTTPClient(options)
	if err != nil {
		return nil, err
	}
	return &AppMeshServerHttpContext{client: httpClient}, nil
}
func newHTTPContextWithRequester(options Option, r Requester) (*AppMeshServerHttpContext, error) {
	httpClient, err := newHTTPClientWithRequester(options, r)
	if err != nil {
		return nil, err
	}
	return &AppMeshServerHttpContext{client: httpClient}, nil
}

// getRuntimeEnv reads and validates required runtime environment variables.
func (r *AppMeshServerHttpContext) getRuntimeEnv() (key, appName string, err error) {
	key = os.Getenv("APP_MESH_PROCESS_KEY")
	appName = os.Getenv("APP_MESH_APPLICATION_NAME")

	if key == "" {
		return "", "", errors.New("missing environment variable: APP_MESH_PROCESS_KEY. This must be set by App Mesh service")
	}
	if appName == "" {
		return "", "", errors.New("missing environment variable: APP_MESH_APPLICATION_NAME. This must be set by App Mesh service")
	}
	return key, appName, nil
}

// TaskFetch fetches task data (payload) for the current application process.
// It retries indefinitely until success. If a request fails within 100ms,
// sleeps briefly before retrying; otherwise retries immediately.
func (r *AppMeshServerHttpContext) TaskFetch() (string, error) {
	key, appName, err := r.getRuntimeEnv()
	if err != nil {
		return "", err
	}

	path := "/appmesh/app/" + appName + "/task"
	query := url.Values{}
	query.Set("process_key", key)

	const retryDelay = 100 * time.Millisecond

	for {
		attemptStart := time.Now()
		status, body, _, err := r.client.get(path, query, nil)
		if err != nil {
			log.Printf("task_fetch request failed: %v, retrying...", err)
		} else if status == http.StatusOK {
			return string(body), nil
		} else {
			log.Printf("task_fetch failed with status %d: %s, retrying...", status, string(body))
		}

		if remaining := retryDelay - time.Since(attemptStart); remaining > 0 {
			time.Sleep(remaining)
		}
	}
}

// TaskReturn sends the processed result bytes back to the original invoking client via App Mesh.
func (r *AppMeshServerHttpContext) TaskReturn(result string) error {
	processKey, appName, err := r.getRuntimeEnv()
	if err != nil {
		return err
	}

	path := "/appmesh/app/" + appName + "/task"
	query := url.Values{}
	query.Set("process_key", processKey)
	headers := map[string]string{"Content-Type": "text/plain"}

	status, body, err := r.client.put(path, query, headers, []byte(result))
	if err != nil {
		return err
	}

	if status != http.StatusOK {
		return errors.New("task_return failed with status " + http.StatusText(status) + ": " + string(body))
	}

	return nil
}

File: src/sdk/rust/src/server_http.rs
// server_http.rs
// Server SDK implementation for App Mesh (HTTP)

use bytes::Bytes;
use http::StatusCode;
use log::{error, warn};
use std::collections::HashMap;
use std::env;
use std::sync::Arc;
use std::time::Duration;
use std::time::Instant;

use crate::client_http::AppMeshClient;
use crate::error::AppMeshError;
use crate::response_ext::ResponseExt;

/// Server-side helper for applications to fetch tasks and return results.
pub struct AppMeshServer {
    client: Arc<AppMeshClient>,
    retry_delay: Duration,
}

impl AppMeshServer {
    /// Create a new server-side task context backed by the HTTP client.
    pub fn new(
        rest_url: Option<String>,
        ssl_verify: Option<String>,
        ssl_client_cert: Option<(String, String)>,
    ) -> Result<Arc<Self>, AppMeshError> {
        let client = AppMeshClient::new(rest_url, ssl_verify, ssl_client_cert, None, None, false)?;
        Ok(AppMeshServer::with_client(client))
    }

    fn get_runtime_env() -> Result<(String, String), AppMeshError> {
        let process_key = env::var("APP_MESH_PROCESS_KEY").map_err(|_| {
            AppMeshError::ConfigurationError(
                "Missing environment variable: APP_MESH_PROCESS_KEY. This must be set by App Mesh service.".into(),
            )
        })?;
        let app_name = env::var("APP_MESH_APPLICATION_NAME").map_err(|_| {
            AppMeshError::ConfigurationError(
                "Missing environment variable: APP_MESH_APPLICATION_NAME. This must be set by App Mesh service."
                    .into(),
            )
        })?;
        Ok((process_key, app_name))
    }

    /// Create `AppMeshServer` with a pre-built client.
    pub fn with_client(client: Arc<AppMeshClient>) -> Arc<Self> {
        Arc::new(Self { client, retry_delay: Duration::from_millis(100) })
    }

    /// Fetch a task payload for the current application process.
    ///
    /// Retries indefinitely until successful. If a request fails within 100ms,
    /// sleeps briefly before retrying; otherwise retries immediately.
    pub async fn task_fetch(&self) -> Result<Bytes, AppMeshError> {
        use reqwest::Method;
        use tokio::time::sleep;

        let (pkey, app_name) = Self::get_runtime_env()?;
        let path = format!("/appmesh/app/{}/task", app_name);

        let mut query = HashMap::new();
        query.insert("process_key".to_string(), pkey);

        let mut attempts: u32 = 0;
        loop {
            let attempt_start = Instant::now();
            match self.client.raw_request(Method::GET, &path, None, None, Some(query.clone()), false).await {
                Ok(resp) => {
                    if resp.status() == StatusCode::OK {
                        return Ok(resp.into_bytes());
                    }
                    let status = resp.status();
                    let text = resp.text()?;
                    warn!("task_fetch attempt {} failed with status {}: retrying...", attempts + 1, status);
                    let _ = (status, text);
                }
                Err(e) => {
                    warn!("task_fetch attempt {} request failed: {}: retrying...", attempts + 1, e);
                }
            }

            attempts += 1;
            if let Some(remaining) = self.retry_delay.checked_sub(attempt_start.elapsed()) {
                sleep(remaining).await;
            }
        }
    }

    /// Return the processed result bytes back to App Mesh so the invoker can receive them.
    pub async fn task_return(&self, result: &[u8]) -> Result<(), AppMeshError> {
        use reqwest::Method;

        let (pkey, app_name) = Self::get_runtime_env()?;
        let path = format!("/appmesh/app/{}/task", app_name);

        let mut query = HashMap::new();
        query.insert("process_key".to_string(), pkey);

        let resp =
            self.client.raw_request(Method::PUT, &path, Some(result), None, Some(query), false).await?;

        let status = resp.status();
        if status != StatusCode::OK {
            let text = resp.text()?;
            error!("task_return failed with status {}: {}", status, text);
            return Err(AppMeshError::RequestFailed { status, message: text });
        }
        Ok(())
    }
}
